src/target/riscv/batch.h

```diff
@@ -14,6 +14,67 @@ enum riscv_scan_type {
 	RISCV_SCAN_TYPE_WRITE,
 };
 
+/* These types are used to specify how many JTAG RTI cycles to add after a
+ * scan.
+ */
+enum riscv_scan_delay_class {
+	/* Delay needed for accessing debug module registers: */
+	RISCV_DELAY_BASE,
+	/* Delay for execution of an abstract command: */
+	RISCV_DELAY_ABSTRACT_COMMAND,
+	/* Delay for System Bus read operation: */
+	RISCV_DELAY_SYSBUS_READ,
+	/* Delay for System Bus write operation: */
+	RISCV_DELAY_SYSBUS_WRITE,
+};
+
+struct riscv_scan_delays {
+	/* The purpose of these delays is to be passed to "jtag_add_runtest()",
+	 * which accepts an "int".
+	 * Therefore, they should be no greater then "INT_MAX".
+	 */
+	unsigned int base_delay;
+	unsigned int ac_delay;
+	unsigned int sb_read_delay;
+	unsigned int sb_write_delay;
+};
+
+static inline unsigned int riscv_scan_get_delay(struct riscv_scan_delays delays,
+		enum riscv_scan_delay_class delay_class)
+{
+	switch (delay_class) {
+	case RISCV_DELAY_BASE:
+		return delays.base_delay;
+	case RISCV_DELAY_ABSTRACT_COMMAND:
+		return delays.ac_delay;
+	case RISCV_DELAY_SYSBUS_READ:
+		return delays.sb_read_delay;
+	case RISCV_DELAY_SYSBUS_WRITE:
+		return delays.sb_write_delay;
+	}
+	return 0;
+}
+
+static inline void riscv_scan_set_delay(struct riscv_scan_delays *delays,
+		enum riscv_scan_delay_class delay_class, unsigned int delay)
+{
+	assert(delay <= INT_MAX);
```

Comment:
The exponential back-off for the delays `next_delay = curr_delay + 1 + round_down(curr_delay * 0.1)` takes currently approx 200 rounds to exceed INT_MAX on 32 bits. 

So it is something that is not that impossible to reach.

For that reason, I would recommend to replace the assertion with saturation. The max value should be chosen so that the int don't overflow on the next increment. For example:

```suggestion
	/* Limit the delay to a value that is guaranteed to not overflow the int datatype on the next increment. */
	if (delay >= RISCV_MAX_DELAY) {  
		delay = RISCV_MAX_DELAY;
		// RISCV_MAX_DELAY can be e.g. 1000000000,
		// which is approx half of INT_MAX on 32 bits
	}	
```

Reply:
This is a valid concern. I've tried to handle it in #1087. Please, take a look

Reply:
Thank you. I have posted my comments under #1087. (And this thread can be closed.)